OpenWeather/App/Search/SearchViewModel.swift

```diff
@@ -0,0 +1,42 @@
+//
+//  OpenWeather template generated by OpenBytes on 22/03/2023.
+//
+// Created by Ahmed Shendy.
+//  SearchViewModel.swift
+//
+
+import Foundation
+import CoreLocation
+
+final class SearchViewModel: ObservableObject, ErrorHandling {
+    private var locationProviding: LocationProviding
+    private var task: Task<Void, Never>?
+
+    var errorHandler: ErrorHandler
+
+    @Published var result: [DeviceLocation] = []
+
+    init(
+        locationProviding: LocationProviding,
+        errorHandler: ErrorHandler = ErrorHandler(
+            plugins: [
+                ToastErrorPlugin()
+            ]
+        )
+    ) {
+        self.locationProviding = locationProviding
+        self.errorHandler = errorHandler
+    }
+
+    func getLocations(query: String) {
+        task?.cancel()
+
+        task = Task {
+            let result = try? await locationProviding.locations(for: query)
+
+            await MainActor.run {
+                self.result = result ?? []
+            }
```

Comment:
I think we might need to add some cancellation check... I don't believe tasks just cancel on their own..

We can use either:
```swift
try Task.checkCancellation()

// From Fork
try await Task.withCheckedCancellation {
    // async
}
```

Fork's `withCheckedCancellation` will check cancellation before the closure and after the closure. Either is fine to use.

Reply:
Great question on if we should handle the cancels, I recall coming across this when I was looking at Tasks but, I think it's important that we handle cancellations when we want to, or at least make the request.

> Swift’s tasks use cooperative cancellation, which means that although we can tell a task to stop work, the task itself is free to completely ignore that instruction and carry on for as long as it wants. This is a feature rather than a bug: if cancelling a task made it stop work immediately, the task might leave your program in an inconsistent state.